src/QAT/qat/purr/compiler/execution.py

```diff
@@ -428,6 +428,19 @@ def create_duration_timeline(self, package: QatFile):
             if all(isinstance(pos_data.instruction, Delay) for pos_data in timeline):
                 del results[key]
 
+        if final_positions := [final_position[-1].end for final_position in list(results.values())]:
+            end_point = max(final_positions)
+
+            for qubit in self.model.qubits:
+                if freq_channel := qubit.pulse_channels.get("freq_shift", None):
```

Comment:
I forget what this is key'd on, but `ChannelType.freq_shift.name()` or similar will be better.